src/library/browse/browsethread.cpp

```diff
@@ -14,80 +14,109 @@ namespace {
-    m_bStopThread = false;
-    m_model_observer = nullptr;
-    //start Thread
+BrowseThread::BrowseThread(QObject* parent)
+        : QThread{parent}, m_modelObserver{}, m_bRun{} {
+    // Start thread
     start(QThread::LowPriority);
-
+    qDebug() << "Wait to start browser background thread";
+    // Wait until the thread is running
+    m_mutex.lock();
+    while (!m_bRun) {
+        m_condition.wait(&m_mutex);
+    }
+    m_mutex.unlock();
+    qDebug() << "Browser background thread started";
 }
 
 BrowseThread::~BrowseThread() {
-    qDebug() << "Wait to finish browser background thread";
-    m_bStopThread = true;
-    //wake up thread since it might wait for user input
-    m_locationUpdated.wakeAll();
-    //Wait until thread terminated
-    //terminate();
+    qDebug() << "Wait to terminate browser background thread";
+    // Inform the thread we want it to terminate
+    m_mutex.lock();
+    m_bRun = false;
+    m_condition.wakeOne();
+    m_mutex.unlock();
+    // Wait until thread terminated
     wait();
-    qDebug() << "Browser background thread terminated!";
+    qDebug() << "Browser background thread terminated";
 }
 
 // static
 BrowseThreadPointer BrowseThread::getInstanceRef() {
-    BrowseThreadPointer strong = m_weakInstanceRef.toStrongRef();
+    // We create a single BrowseThread instead which is used by multiple
+    // BrowseTableModel instances. We store a weakpointer so when all
+    // BrowseTableModel have been destroyed, so it the BrowseThread.
+
+    static QWeakPointer<BrowseThread> s_weakInstanceRef;
+    static QMutex s_mutex;
+
+    s_mutex.lock();
+    BrowseThreadPointer strong = s_weakInstanceRef.toStrongRef();
     if (!strong) {
-        s_Mutex.lock();
-        strong = m_weakInstanceRef.toStrongRef();
-        if (!strong) {
-            strong = BrowseThreadPointer(new BrowseThread());
-            m_weakInstanceRef = strong.toWeakRef();
-        }
-        s_Mutex.unlock();
+        strong = BrowseThreadPointer(new BrowseThread());
+        s_weakInstanceRef = strong.toWeakRef();
     }
+    s_mutex.unlock();
+
     return strong;
 }
 
-void BrowseThread::executePopulation(mixxx::FileAccess path, BrowseTableModel* client) {
-    m_path_mutex.lock();
+void BrowseThread::requestPopulateModel(mixxx::FileAccess path, BrowseTableModel* modelObserver) {
+    // Inform the thread to populate a new path.
+    // Note: if another request is currently being processed, this will replace it.
+
+    qDebug() << "Request populate model" << path.info() << modelObserver;
+    m_mutex.lock();
     m_path = std::move(path);
-    m_model_observer = client;
-    m_path_mutex.unlock();
-    m_locationUpdated.wakeAll();
+    m_modelObserver = modelObserver;
+    m_condition.wakeOne();
+    m_mutex.unlock();
 }
 
 void BrowseThread::run() {
     QThread::currentThread()->setObjectName("BrowseThread");
+
+    // Inform the constructor the thread started
     m_mutex.lock();
+    m_bRun = true;
+    m_condition.wakeOne();
```

Comment:
wakeOne() and wait() below is confusing, Let's use a second sync method.

Reply:
I put the request in a separate method.